### ISSUE TITLE
ng-template gets rendered twice during hydration (when anchor comment node is not projected)

### ISSUE THREAD

**Issue by michael-hein** [angular/angular#50543]
### Which @angular/* package(s) are the source of the bug?

platform-browser

### Is this a regression?

No

### Description

when hydration is enabled and a component uses `ngTemplateOutlet` to render a template passed from a parent e.g. via `ContentChild` then its getting rendered twice for the duration of the hydration.

So it looks the like this if 'abc' is the template content
received from server
`abc`

during hydration
`abc abc` <-- this looks sometimes a bit ugly

after hydration
`abc`

So is it possible that the passed template gets just once rendered during the hydration process?

The relevant code parts for reprodruction:
```
<app-shell>
  <ng-template #customTemplate>
    <p>template</p>
  </ng-template>
</app-shell>
```

```
@Component({
  selector: 'app-shell',
  template: `<ng-container
    *ngIf="customTemplate"
    [ngTemplateOutlet]="customTemplate"
  ></ng-container>`,
  styles: [],
})
export class ShellComponent {
  @ContentChild('customTemplate', { static: true })
  customTemplate: TemplateRef<any> | null = null;
}

```

### Please provide a link to a minimal reproduction of the bug

https://github.com/michael-hein/hydration-ng-template

### Please provide the exception or error you saw

```true
N/A
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 16.0.4        
Node: 18.16.0
Package Manager: npm 8.19.3
OS: win32 x64

Angular: 16.0.4
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1600.4
@angular-devkit/build-angular   16.0.4
@angular-devkit/core            16.0.4
@angular-devkit/schematics      16.0.4
@nguniversal/builders           16.0.2
@nguniversal/express-engine     16.0.2
@schematics/angular             16.0.4
rxjs                            7.8.1
typescript                      4.9.5
```

**Comment by dylhunn** [comment]
(@jessicajaniuk and @AndrewKushnir for visibility.)

**Comment by michael-hein** [comment]
thanks for the fast response and for looking into it. I really appreciate it :)

Yes the `setTimeout` was only to better see it, that was rendered twice.

With the `<ng-content />` tag it works perfectly. Thank you!
```
<app-shell>
    <ng-template #customTemplate>
      <p>template</p>
    </ng-template>
  </app-shell>
```

```
@Component({
  selector: 'app-shell',
  template: `<ng-content /><ng-container
      *ngIf="customTemplate"
      [ngTemplateOutlet]="customTemplate"
    ></ng-container>`,
  styles: [],
})
export class ShellComponent {
  @ContentChild('customTemplate', { static: true })
  customTemplate: TemplateRef<any> | null = null;
}
```

**Comment by CarlosTorrecillas** [comment]
@AndrewKushnir I have tried the `<ng-content />` without success. Perhaps I am not doing something wrong. In my scenario I have

```
<ng-template #template>
    <ng-container *ngIf="googleBusiness">
        <li>
            Google Business info 1
        </li>
        <li>
            Google Business info 2
        </li>
    </ng-container>
</ng-template>
```

And my component looks like:

```
export class GoogleBusinessComponent implements OnInit {
    @Input() googleBusiness: IGoogleBusiness;
    @ViewChild('template', { static: true }) template: TemplateRef<unknown>;

    constructor(private viewContainerRef: ViewContainerRef) { }
      
    ngOnInit() {
        this.viewContainerRef.createEmbeddedView(this.template);
        this.viewContainerRef.element.nativeElement.remove();
    }
}
```
Essentially I want to remove the component tag from being rendered since that component adds two <li> items to an existing list. I have added the `<ng-content />` tag inside the `ng-template`, and outside. Do I have to do anything else to prevent the component being rendered twice?

**Comment by AndrewKushnir** [comment]
@CarlosTorrecillas the problem that you describe might be different from the one reported here. Could you please try to update your app to v16.1.1 (most recent at this moment) and test the behavior once again? If the problem still exists, please create a minimal repro (in a form of a GitHub repository with a new `ng new` app) and submit a new ticket, we'll take a look. Thank you.

**Comment by CarlosTorrecillas** [comment]
Hi @AndrewKushnir , I do have version 16.1.1 and tried different things but the error seems to be there. I will try to get a sample repo with the issue tomorrow and will submit the defect accordingly

**Comment by CarlosTorrecillas** [comment]
Good morning @AndrewKushnir , I have created https://github.com/angular/universal/issues/3179 with all the repro steps in it. Let me know if you need further info. Thanks!

**Comment by jwittekind** [comment]
+1 
Any news on this topic?
I'm using angular 17 and in my case i'm using content directives:

````
        <ng-container *ngTemplateOutlet="content?.templateRef || defaultContentTemplate; context: context" />
        <ng-template #defaultContentTemplate>
                <!-- This content is duplicate when rendered by ssr  -->
        </ng-template>        
````

**Comment by JeanMeche** [comment]
@jwittekind Could you provide a working repro on a github repo  ? This basic example without its context doesn't seem to reproduce the issue.

**Comment by DjordjeDursun** [comment]
I had the same problem, and it was related to change detection, i fixed it by adding "changeDetection: ChangeDetectionStrategy.OnPush,", to "@Component({})" for all of my components

**Comment by DjordjeDursun** [comment]
@michael-hein

**Comment by jwittekind** [comment]
> I had the same problem, and it was related to change detection, i fixed it by adding "changeDetection: ChangeDetectionStrategy.OnPush,", to "@component({})" for all of my components

I can not confirm that adding "changeDetection: ChangeDetectionStrategy.OnPush," solves the problem for me

**Comment by Roman-Simik** [comment]
Any update on this?

**Comment by AndrewKushnir** [comment]
> Any update on this?

The root cause of the initially reported issue is described in https://github.com/angular/angular/issues/50543#issuecomment-1572900259 and the solution is to make sure that the `<ng-content />` gets used in a component that does content projection. If the mentioned solution doesn't work for your scenarios, there is a chance that the root cause is different and we'd need a minimal repro (in a form of a GitHub repository with a new `ng new` app and minimal amount of code to repro the issue).

**Comment by Tayuelo** [comment]
Hi @AndrewKushnir, in my case I'm using PrimeNG and I'm facing the same issue that's discussed here, I believe the root cause is what you explained before but I cannot add the `<ng-content />` since this is a third-party library.

This is how I'm using the TabMenuComponent from PrimeNG:

      <p-tabMenu
          #tabMenu
          [model]="tabs()"
          [activeItem]="activeTab()">
          <ng-template pTemplate="item" let-item>
            <a [routerLink]="item.route" class="p-menuitem-link">
              <span class="ml-2">
                {{ item.label }}
              </span>
            </a>
          </ng-template>
      </p-tabMenu>

This is the code from PrimeNG [TabMenu](https://github.com/primefaces/primeng/blob/e624bb5b706eefc384dc680d0d60fb3fbb79987c/src/app/components/tabmenu/tabmenu.ts#L117C42-L117C43)

What would you suggest in this scenario?

**Comment by Roman-Simik** [comment]
@AndrewKushnir yes, thanks, that described solution above works, but my question was towards Angular devs, if they are planning to fix it, so we don't need to handle it manually by ourselfs by adding the <ng-content/> tag

**Comment by jwittekind** [comment]
+1 I also experience this issue with `<ng-content/>`, `<ng-template/>` or custom templateRefs using `@ContentChild`.
Please fix this issue

**Comment by tsteuwer-accesso** [comment]
We are building a large ecommerce application using PrimeNG as well and are now running into this issue when we're using `<ng-template />`. Is there any plans from the Angular team to address this anytime soon? It's been almost 2 years.

**Comment by spowka** [comment]
Is there any update or planned milestone for improving hydration behavior in these scenarios? This issue affects some common patterns in SSR-heavy apps where conditional rendering is handled within `ng-template`.

**Comment by thePunderWoman** [comment]
Hey everyone. We were able to figure out a fix for this and it will be released with the patch that goes out next week Wednesday. Thanks for your patience!

**Comment by tsteuwer-accesso** [comment]
Thank you @thePunderWoman !!!!
